Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,640 @@
+// normalized via NSDecimalNormalize (the exponents of the both decimals will be the same)
+static void _InternalDecimalCopy(NSDecimal* destination, NSDecimalInternal* source);
+static NSComparisonResult _compareMantissa(NSDecimal* left, NSDecimal* right);
+
+static NSCalculationError _handleDifferenceOfSignAddition(NSDecimal* result,
+                                                          NSDecimal* left,
+                                                          NSDecimal* right,
+                                                          NSRoundingMode roundingMode);
+
+static NSCalculationError _subtraction(NSDecimal* result, const NSDecimal* left, const NSDecimal* right);
+static NSCalculationError _mantissaAddition(NSDecimalInternal* result, NSDecimalInternal* augend, unsigned short addend);
+static NSCalculationError _mantissaMultiplication(NSDecimalInternal* result, NSDecimalInternal* multiplicand, unsigned short multiplier);
+static unsigned int _obtainLength(unsigned short* mantissa, int length);
+static NSCalculationError _mantissaDivision(NSDecimalInternal* result, NSDecimalInternal* dividend, unsigned short divisor);
+static NSCalculationError _compactMantissa(NSDecimalInternal* iDecimal, NSRoundingMode roundingMode);
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCopy(NSDecimal* destination, const NSDecimal* source) {
+    if (!source || !destination) {
+        return;
+    }
+
+    destination->_isNegative = source->_isNegative;
+    destination->_exponent = source->_exponent;
+    destination->_length = source->_length;
+    destination->_isCompact = source->_isCompact;
+    destination->_reserved = source->_reserved;
+    memcpy(&destination->_mantissa, &source->_mantissa, NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+/**
+ @Status Interoperable
+*/
+BOOL NSDecimalIsNotANumber(const NSDecimal* dcm) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, dcm != NULL);
+    return !dcm->_length && IS_NEGATIVE(dcm);
+}
+
+/**
+ @Status Interoperable
+*/
+NSComparisonResult NSDecimalCompare(const NSDecimal* leftOperand, const NSDecimal* rightOperand) {
+    if (leftOperand == rightOperand) {
+        return NSOrderedSame;
+    }
+
+    THROW_NS_IF_FALSE(E_INVALIDARG, (leftOperand != NULL) && (rightOperand != NULL));
+
+    // not a number
+    if (NSDecimalIsNotANumber(leftOperand)) {
+        if (NSDecimalIsNotANumber(rightOperand)) {
+            return NSOrderedSame;
+        }
+        return NSOrderedAscending;
+    } else if (NSDecimalIsNotANumber(rightOperand)) {
+        return NSOrderedDescending;
+    }
+
+    NSDecimal left, right;
+    NSDecimalCopy(&left, leftOperand);
+    NSDecimalCopy(&right, rightOperand);
+    NSDecimalNormalize(&left, &right, NSRoundPlain);
+
+    if (!left._isNegative && right._isNegative) {
+        return NSOrderedDescending;
+    }
+
+    if (left._isNegative && !right._isNegative) {
+        return NSOrderedAscending;
+    }
+
+    // same signs
```

Comment:
from here to the end of the section can just be _compareMantissa, then return the opposite if right._isNegative